hypervisor/src/mshv/mod.rs

```diff
@@ -611,6 +617,392 @@ impl cpu::Vcpu for MshvVcpu {
+                                    };
+
+                                    let port =
+                                        // SAFETY: Accessing a union element from bindgen generated bindings.
+                                        unsafe { port_info.__bindgen_anon_1.intercepted_port() };
+                                    let mut len = 4;
+                                    // SAFETY: Accessing a union element from bindgen generated bindings.
+                                    unsafe {
+                                        if port_info.__bindgen_anon_1.operand_size_16bit() == 1 {
+                                            len = 2;
+                                        } else if port_info.__bindgen_anon_1.operand_size_8bit()
+                                            == 1
+                                        {
+                                            len = 1;
+                                        }
+                                    }
+                                    let is_write =
+                                        // SAFETY: Accessing a union element from bindgen generated bindings.
+                                        unsafe { port_info.__bindgen_anon_1.access_type() == 0 };
+                                    let mut rax_gpa: mshv_read_write_gpa =
+                                        mshv_bindings::mshv_read_write_gpa {
+                                            base_gpa: ghcb_gpa + GHCB_RAX_OFFSET,
+                                            byte_count: std::mem::size_of::<u64>() as u32,
+                                            ..Default::default()
+                                        };
+                                    self.fd
+                                        .gpa_read(&mut rax_gpa)
+                                        .map_err(|e| cpu::HypervisorCpuError::GpaRead(e.into()))?;
+
+                                    let mut bytes = [0u8; 8];
+                                    bytes.copy_from_slice(&rax_gpa.data[0..8]);
+
+                                    let rax = u64::from_le_bytes(bytes);
+                                    let data = (rax as u32).to_le_bytes();
+                                    if is_write {
+                                        if let Some(vm_ops) = &self.vm_ops {
+                                            vm_ops.pio_write(port.into(), &data[0..len]).map_err(
+                                                |e| cpu::HypervisorCpuError::RunVcpu(e.into()),
+                                            )?;
+                                        }
+                                    } else {
+                                        let mut data: [u8; 4] = [0; 4];
+                                        if let Some(vm_ops) = &self.vm_ops {
+                                            vm_ops
+                                                .pio_read(port.into(), &mut data[0..len])
```

Comment:
Again, I might be wrong here but can we not we just copy directly into `rax_gpa.data`?

Reply:
Correct fixed in the new revision